### PR TITLE
Fix/lambda errors

### PR DESCRIPTION
**Pull request by szymonm**
Relates: #8028 

Ideas for next steps, @electrum @kokosing @findepi ?

### CODE REVIEW

**Comment by kokosing** [comment]
Make sure you have read all the https://github.com/prestodb/presto#developers, especially git commit messages.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    public static List<IndependentTypeSignatureProvider> fromTypes(List<? extends Type> types)
+    {
+        return types.stream()
+                .map(Type::getTypeSignature)
+                .map(IndependentTypeSignatureProvider::new)
+                .collect(toImmutableList());
+    }
+
+    public static List<IndependentTypeSignatureProvider> fromTypeSignatures(List<? extends TypeSignature> typeSignatures)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java]: put static methods at the top

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class IndependentTypeSignatureProvider implements TypeSignatureProvider
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java]: I do not like the world `independent` here. It is just plain, regular, basic, the most common type signature provider, while independent means it is very special.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+{
+    private final TypeSignature typeSignature;
+
+    public IndependentTypeSignatureProvider(TypeSignature typeSignature)
+    {
+        this.typeSignature = typeSignature;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java]: use `requireNonNull` aka `rnn` (create a live template for such kind of stuff)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    @Override
+    public boolean hasDependency()
+    {
+        return false;
+    }
+
+    @Override
+    public TypeSignature getTypeSignature(List<Type> boundTypeParameters)
+    {
+        checkState(boundTypeParameters.isEmpty(), "Independent type signature doesn't take any type parameters.");
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java]: did you mean `checkArgument`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -354,6 +355,9 @@ protected Type visitSymbolReference(SymbolReference node, StackableAstVisitorCon
                 }
             }
             Type type = symbolTypes.get(Symbol.from(node));
+            if (type == null) {
+                throw new SemanticException(UNKNOWN_IDENTIFIER, node, "Unknown identifier: %s", node);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: I think you should use `MISSING_ATTRIBUTE`

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayReduceFunction.java [hunk 6]
```diff
@@ -110,4 +110,17 @@ public void testErrorMessageOnUnknownIdentifier()
...
+    public void testErrorMessageOnWrongNumberOfArguments()
+    {
+        assertInvalidFunction("reduce(ARRAY [1, 2], 0, (s, x) -> s + x, s -> s, x -> x)", FUNCTION_NOT_FOUND,
+                "Unexpected parameters for function reduce. Expected: reduce(array(T), S, function(S,T,S), function(S,R)) for type(s): T, S, R, (wrong number of arguments: expected 4 got 5)");
+    }
+
+    @Test
+    public void testErrorMessageOnInvalidLambdaBody()
+    {
+        assertInvalidFunction("reduce(ARRAY [1, 2], 0, (s, x) -> concat(s, x), s -> s)", FUNCTION_NOT_FOUND);
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayReduceFunction.java]: what is the message here?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java (hunk 3 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java (hunk 2 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java]: Done.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java (hunk 4 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java (hunk 1 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/IndependentTypeSignatureProvider.java]: Done

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java (hunk 5 above)
Comment (szymonm) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: Ok

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayReduceFunction.java (hunk 6 above)
Comment (szymonm) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayReduceFunction.java]: Added. Though, it's not the part I'm very proud of. Any suggestions for a message in this case?

**Comment by szymonm** [comment]
@kokosing can you look again on this?

**Comment by szymonm** [comment]
@haozhun Do you have any thoughts on what we should do next with errors in lambdas?

**Comment by haozhun** [comment]
Please do not separate the two TypeSignatureProvider. It was a careful and deliberate decision

# Your refactor changed semantics

In your refactor, you likely noticed that there is not a single place where we did a branch before invoking `typeSignatureProvider`. 

```
if (is lambda) {
  return typeSignatureProvider.getTypeSignature(inputs);
}
else {
  return typeSignatureProvider.getTypeSignature();
}
```

The semantics of the assertion also changed after your commit. You did not assert that list is not empty in `LambdaTypeSignatureProvider.getTypeSignature`. You can't assert that because that will fail.

By doing this, two distinct scenarios became the same.

# Interface doesn't help

* Once you revert the semantic change, you will notice that there isn't really shared methods. There basically 3 methods, and 2 of them are only available in one of the implementation.
* Whenever those 2 `getTypeSignature` are used, the context is always clear which kind it is.
* The class is simple and short.
* If and when we need a nother implementation, we could re-evaluate this then

# Random thoughts

* The 2 static methods (fromTypes and fromTypeSignatures) should be in the interface even if you separate TypeSignatureProvider into 2 classes.
* Some of the naming is wrong. Maybe `hasDependency` should be renamed to `isFunctionType`. Maybe `getTypeSignature` should have 2 different names.

**Pull request closed without merging**